spec/asyncapi.md

```diff
@@ -139,6 +139,21 @@ This is applicable for `$ref` fields in the specification as follows from the [J
 
 By convention, the AsyncAPI Specification (A2S) file is named `asyncapi.json` or `asyncapi.yaml`.
 
+### <a name="traitInheritance"></a>Trait Inheritance
+
+When traits ([Operation Trait Object](#operationTraitObject) and [Message Trait Object](#messageTraitObject)) are defined, they MUST be merged into the target object according to the following rules:
+
+* Traits are merged with a recursive merge algorithm, as defined in the [JSON Merge Patch](https://tools.ietf.org/html/rfc7386) RFC. To summarize the main merge rules as a merge from source object into a target object:
```

Comment:
I'm confused now. If this is as defined in JSON Merge Patch, why do we have custom rules below? I'm also having a hard time figuring out how this would work.  Maybe we could have some examples along with the explanation?

Reply:
Yes, I'm also unsure about that.

So the RFC already states the merge rules, but it also covers many other aspects that are not relevant to this trait inheritance at all. If you get the rules right, you could just state a few rules here. I still thinks its valuable to say that this is RFC-7386 conformant, so you can reuse existing libraries.

In the linked issue, I've referenced some examples and code. I wasn't sure if this should be part of the specification itself? But having some examples and code is always helful if you can link to it.

See https://jsfiddle.net/FannonF/yLsonr57/27/

Reply:
@fmvilas Btw. some of those rules are still necessary because they are not part of the JSON Merge Patch RFC.

The following rules are specific to AsyncAPI itself and have been missing (undefined) previously:

* Before applying merges, JSON Schema `$ref` MUST be dereferenced
* The [Message Object](#messageObject) and the [Operation Object](#operationObject) that the trait is applied to is the most specific and is therefore never overwritten, only extended.

The following rule is AsyncAPI specific but was already there before I think:
* The order of the traits defines their specificity. Subsequent traits in the trait array are more specific.

Reply:
Maybe we should just explicitly state the extension to the RFC, that way we ensure there is less interpretation involved.

Something similar to:

```md
The traits ([Operation Trait Object](#operationTraitObject) and [Message Trait Object](#messageTraitObject)) are defined using [JSON Merge Patch](https://tools.ietf.org/html/rfc7386). In extension to these merge rules, the following applied:
- Any references must be resolved before applying traits. 
- The Message Object and the Operation Object that the trait is applied to is the most specific and is therefore never overwritten, only extended.
- The order of the traits defines their specificity. Subsequent traits in the trait array are more specific.
...
```

Reply:
For your coding examples, do you mind writing a comment in this issue: https://github.com/asyncapi/parser-api/issues/36

This way we can include your examples as documentation and explanation of how to achieve the correct behavior in tooling 🙂

Reply:
Hi Jonas,

sure, I can do that! However, we were not completely agreeing on what the future approach should be. 
This code example was my take on how I would do it ;)

Best,
Simon

Reply:
Ahh yea, my bad, I saw this PR as an attempt to clarify the current behavior. Gonna follow up in https://github.com/asyncapi/spec/issues/557